docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,803 @@
+`RealmKeyRotationCertificate` (same author, timestamp, realm_id and key_index)
+to ensure this message is legit.
+
+### New certificate: `realm_key_rotation_certificate`
+
+```json5
+{
+    "label": "RealmKeyRotationCertificate",
+    "type": "realm_key_rotation_certificate",
+    "other_fields": [
+        {
+            "name": "author",
+            "type": "DeviceID"
+        },
+        {
+            "name": "timestamp",
+            "type": "DateTime"
+        },
+        {
+            "name": "realm_id",
+            "type": "VlobID"
+        },
+        {
+            "name": "key_index",
+            "type": "Index"
+        }
+    ]
+}
+```
+
+When processing this certificate (in addition to regular access checks), the
+client must make sure the `key_index` field strictly follow the value of the
+previous key rotation certificate for this realm. (i.e. we reject the
+certificate if there is a gap in key indexes)
+
+## 5 Changes: Server-side
+
+### 5.1 Change HTTP `Server` header
+
+Currently the HTTP `Server` value is `parsec`, we should change it to `parsec/3` to
+simplify the detection of an incompatible server (see
+[7.1 - No Parsec v2 to v3 auto-update](#71-no-parsec-v2-to-v3-auto-update) ).
+
+### 5.2 Datamodel
+
+The main issue with data migration is handling realm currently under maintenance given
+it asks the question of which key to keep:
+
+- Keeping only the last key doesn't work as some data are not encrypted for it
+- Keeping the n-1 keep only makes migration complicated on client-side (the server
+  no longer has concept of realm under migration, however the client has to find
+  a way to know that. On top of that the migration process requires accessing
+  older version of the user manifest which add complexity)
+- Keeping both n-1 and n adds basically the same issues with client-side
+
+So we go simple on this and only allow server migration if all the realms have
+finished they maintenance, or if a realm doesn't contain any data (see below).
+
+To avoid being blocked by a workspace forever in maintenance (e.g. workspace owned by
+a single user that got revoked while it was re-encrypting it), we also provide a cli for
+removing realm vlobs (cannot remove a realm given it is related to certificates) or
+organization form the database.
+
+### 5.3 Server data migration
+
+Drop tables:
+
+- `vlob_encryption_revision`
+- `message`
+
+Table `vlob_atom`:
+
+- Remove all rows that contains old encryption revision data (i.e. the table only contains
+  the vlob currently accessible by the users).
+- Remove column `vlob_encryption_revision`
+- Add column `key_index` (`INTEGER NOT NULL`). This defaults to 1 when migrating data.
+
+Table `realm`
+
+- Remove columns `encryption_revision`, `maintenance_started_by`, `maintenance_started_on` and `maintenance_type`
+
+## 6 Changes: Client-side
+
+### 6.1 Certificate Ops
+
+`RealmKeyRotationCertificate` should be checked (only OWNER of the realm can create such
+certificate).
```

Comment:
> Should we include the admin ? since they can remove / add user.
> If we remove an user we should encrypt the new data using a new key ASAP.

This is handled by a new "key rotation manager " that relies on the certificate ops to detect revoked user and then remove they access and do a key rotation.

> That would be cumbersome if when we remove an user, we need to wait for the owner to connect to generate a new set of keys

We have no choice on this: generating a new key can only be done by a user with access to the realm, but ADMIN profile is orthogonal to that.